lib/spack/llnl/util/filesystem.py

```diff
@@ -1180,6 +1187,69 @@ def remove_if_dead_link(path):
         os.unlink(path)
 
 
+class GenericVisitor(object):
+    def __init__(self, callable):
+        pass
+
+    def before_visit_dir(self, root, rel_path, depth):
+        pass
+
+    def after_visit_dir(self, root, rel_path, depth):
+        pass
+
+    def before_visit_symlinked_dir(self, root, rel_path, depth):
+        pass
+
+    def after_visit_symlinked_dir(self, root, rel_path, depth):
+        pass
+
+    def visit_file(self, root, rel_path, depth):
+        pass
+
+    def visit_symlinked_file(self, root, rel_path, depth):
```

Comment:
One thing to clarify in the docstring for this function: will it invoke the callback on the symlink path, or on the path of the target of the symlink?